### PR TITLE
SE-2827: TaxRuleSets notebook

### PR DESCRIPTION
**Pull request by johnd83**
# Pull Request Checklist

- [x] Changes follow the [style guide](https://github.com/finbourne/sample-notebooks/blob/master/docs/FINBOURNE%20notebook%20style%20guide.ipynb)
- [ ] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR
Adding TaxRuleSets notebook and required data.

(Re-opened from PR 600 which seemed to stop working with reviewNB)

### CODE REVIEW

File: examples/use-cases/ibor/Tax Rule Sets.ipynb [hunk 1]
```diff
@@ -0,0 +1,3977 @@
+{
```
Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <ul><li>Add comma after 'In this notebook'</li><li>I would say 'their corresponding dividends' on the second sentence. Just a preference.</li><li>Add comma after 'Finally'</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>I would probably re-write to say 'We then create a function to help upsert equities into LUSID'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Would rephrase this as: 'We fund our portfolio with enough cash such that after we purchase our equities, we're left with £100 and $100 respectively'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Might be worth adding transaction id to the above table. We can then explain how this can be used to delineate multiple dividends in a holdings view.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <ul><li>Add comma after 'payment date'.</li><li>I think you can remove 'as expected' at the end. </li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='52'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Do we need this displayed? Looks like it was displayed once already at the start of section 5.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Maybe just explain that a negative tax rate results in a positive cash inflow and is used for things like tax relief.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='85'/>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Done</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Done</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Done</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Yes - when viewing the notebook as a slideshow this cell allows easy reference to the dividend data on the same slide as the LUSID holdings viewed across multiple dates.</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Done</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: Done

File: examples/use-cases/ibor/Tax Rule Sets.ipynb [hunk 2]
```diff
@@ -0,0 +1,4074 @@
...
```
Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Can this be put in a function?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='101'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Can this be put in a function?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='103'/>

Comment (Tomwp) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Can we move all the delete function calls (rule set, ca source, portfolio, and properties) into a final stand-alone cell. This will make commenting out easier.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='105'/>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: Done

File: examples/use-cases/ibor/Tax Rule Sets.ipynb [hunk 3]
```diff
@@ -0,0 +1,4044 @@
...
```
[the remark this answers is not in the document]
Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: Done

[the remark this answers is not in the document]
Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: Done

File: examples/use-cases/ibor/Tax Rule Sets.ipynb [hunk 4]
```diff
@@ -0,0 +1,4071 @@
...
```
Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Not sure if this is needed, guessing it makes the output neat, but most of the other notebooks rely on simply displaying the df which should probably be fine for this notebook as well. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>We should work with James Leedham to make sure this gets linked to a KB article once one is out. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.            response = tax_rules_api.get_tax_rule_set(scope=scope, code=tax_rule_set_code)
</pre><p>Slight nitpicking, but might be good to make the scope available as a parameter in the top function rather than have it enclosed within this function.</p><p> </p><p>In case someone wants to reuse it means they won't have to do that bit of refactoring here. A bit surprised we don't have an <code>Upsert</code> endpoint to begin with natively in LUSID.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def getRules(inst_property_name, inst_property_value, rate):
</pre><p>Would follow <code>snake_case</code> for the function name here, and also not sure if get_rules is a bit confusing as a method name as it's not actually getting it from LUSID from what I can see? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.                name=f"{inst_property_value}_TaxRule",
</pre><p>Do we need to append the <code>_TaxRule</code>, and if so why?   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>tax amount, or taxable amount? I'm confused to why the tax amount would show up as a receivable. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Noticing now the data is quite a few decimal places, think you'll want to set 2sf at the top of the notebook similar to what's done in the others. </p><p> </p><p>We typically add:</p><p> </p><pre class="ql-syntax" spellcheck="false"># Set DataFrame display formats
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
pd.options.display.float_format = "${:,.2f}".format
display(HTML("&lt;style&gt;.container { width:90% !important; }&lt;/style&gt;"))
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

[the remark this answers is not in the document]
Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Added some more info about what the variables here are to be used for.</p>

[the remark this answers is not in the document]
Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Done</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Have added scope to this function.</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Renamed to build_rules</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Nope - it's just the name so we don't "need" to. I'm easy with/without it, so I've removed "_TaxRule" from the name.</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Correct as "tax amount". There is a row for the gross dividend (holding type 'Accrual') and a second row for the tax due on that dividend (holding type 'Receivable').</p><p> </p><p>A holding of type 'Receivable', per the KB link included in the notebook, is:</p><p><em>"Future cash flow into the portfolio from non-trading activity such subscriptions, redemptions, fees, taxes etc. This label of "Receivable" is used to capture the class of holding types which represent </em><strong><em>both Receivables and Payables.</em></strong><em>"</em></p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: Centre alignment is done to keep data and column headings lined up nicely when viewing the notebook as a workbook. I have reworked this a little to make it clearer.

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: Some floats in this nb should be more precise (e.g. div_amount needs to be shown to 4dp). I have re-worked the formatting throughout the nb and now numbers are shown with thousand separators and appropriate number of dp.

[the remark this answers is not in the document]
Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: The renaming would need doing still because some column names still come out quite long and I'm trying to keep consistent terminology throughout the notebook.

The check for (and adding of) the transaction_id column is so that the rename and subsequent column selection doesn't fail when the column isn't present in the result from LUSID. This happens for dates when there is no unsettled transaction, which in the notebook is dates such as viewing the initial holdings and any date after a dividend's payment date (and before the dividend's ex date).

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: 👍  James and I already in touch about tax rule sets.

File: examples/use-cases/ibor/Tax Rule Sets.ipynb [hunk 5]
```diff
@@ -0,0 +1,4337 @@
...
```
Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    formats = {
</pre><p>Are these needed given you have the global pd. .2f setting on line 3? Or perhaps just needed for the ones where != .2f?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #21.    try:
</pre><p>Not sure you need the try/except. It makes sense to catch specific API exceptions such as "PortfolioAlreadyExists". But in this case if you just run the request, the ApiException will already <code>raise e</code> without this block. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (cscialino) [inline examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #15.        df_holdings = df_holdings[["properties.Instrument/default/Name.value.label_value",
</pre><p>Lines 15-20 probably obsolete. </p><p> </p><p>rename also has the option to <code><strong><em>errors='ignore'</code></em></strong></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/626/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Redundant column selection removed.</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Thanks, yep - changed.</p>

Comment (johnd83) [reply examples/use-cases/ibor/Tax Rule Sets.ipynb:1]: <p>Afraid so. Without them those columns are shown with many more dp than the default still. </p>

**Review by cscialino — APPROVED** [review]
